### PR TITLE
fix(cardinality): Check cardinality with multiple inherited classes (DEV-1189)

### PR DESCRIPTION
**Pull request by irinaschubert**
Resolves DEV-1189, DEV-1206

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: dsp-schema/core/src/main/scala/dsp/schema/domain/SchemaDomain.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    } else {
+      self match {
+        case MustHaveOne  => true
+        case MustHaveSome => that == MayHaveMany
+        case MayHaveOne   => that == MayHaveMany
+        case MayHaveMany  => false
+      }
+    }
+}
+object MustHaveOne extends Cardinality {
```
Comment (subotic) [inline dsp-schema/core/src/main/scala/dsp/schema/domain/SchemaDomain.scala:36]:
```suggestion
object Cardinality {
   final case object MustHaveOne extends Cardinality {
      override val value: String = "1"
   }
}
```

This way, all the different constructors can be found under the `Cardinality` type. Easier discoverability.
If you don't like typing at the call sight, you can always `import Cardinality._`

File: webapi/src/test/scala/org/knora/webapi/responders/v2/ontology/OntologyHelpersSpec.scala [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      // define the classes
+      val mayHaveMany  = freetestOntologyIri.makeEntityIri("PubMayHaveMany").toOntologySchema(InternalSchema)  // 0-n
+      val mustHaveSome = freetestOntologyIri.makeEntityIri("PubMustHaveSome").toOntologySchema(InternalSchema) // 1-n
+      val mayHaveOne   = freetestOntologyIri.makeEntityIri("PubMayHaveOne").toOntologySchema(InternalSchema)   // 0-1
+      val mustHaveOne  = freetestOntologyIri.makeEntityIri("PubMustHaveOne").toOntologySchema(InternalSchema)  // 1
+
+      val hasPublicationDateProperty = freetestOntologyIri.makeEntityIri("hasPublicationDate")
+
+      // define all test cases and the expected results (see comments)
+      val testCases: List[Set[SmartIri]] = List(
```
Comment (subotic) [inline webapi/src/test/scala/org/knora/webapi/responders/v2/ontology/OntologyHelpersSpec.scala:54]: To make it even easier to read/check, you could merge `testCases` and `expectedResults` as tuples per test case:

```
val testCases: List(
  (Set(mayHaveMany), Some(MayHaveMany))
)
```
then do something like this:
```
testCases map { (testCase, expected) =>
  assert(getStrictest(testCase) == expected)
}


def getStrictest(set: Set[SmartIri]) = OntologyHelpers.getStrictestCardinalitiesFromClasses(
          set,
          cacheData
        ).get(hasPublicationDateProperty) match {
            case None       => None
            case Some(card) => Some(card.cardinality)
          }
```

**Review by subotic — APPROVED** [review]
LGTM, thanks!

File: build.sbt [hunk 3]
```diff
@@ -27,7 +27,7 @@ lazy val buildSettings = Seq(
 lazy val rootBaseDir = ThisBuild / baseDirectory
 
 lazy val root: Project = Project(id = "root", file("."))
-  .aggregate(webapi, sipi, shared, valueObjects, userCore, userHandler, userRepo, userInterface)
+  .aggregate(webapi, sipi, shared, userCore, userHandler, userRepo, userInterface)
```
Comment (BalduinLandolt) [inline build.sbt:30]: this is missing the schema-slice and role-slice projects, so those tests are not executed on CI... could you add them?

File: webapi/src/test/scala/org/knora/webapi/messages/v2/responder/ontologymessages/KnoraCardinalityInfoIsStricterThanZSpec.scala [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+object KnoraCardinalityInfoIsStricterThanZSpec extends ZIOSpecDefault {
+
+  def spec = (mustHaveOneTest + mustHaveSomeTest + mayHaveOneTest + mayHaveManyTest)
+
+  val mustHaveOne  = MustHaveOne  // "1"
+  val mustHaveSome = MustHaveSome // "1-n"
+  val mayHaveOne   = MayHaveOne   // "0-1"
+  val mayHaveMany  = MayHaveMany  // "0-n"
+
+  private val mustHaveOneTest = suite("MustHaveOneTest")(
```
Comment (BalduinLandolt) [inline webapi/src/test/scala/org/knora/webapi/messages/v2/responder/ontologymessages/KnoraCardinalityInfoIsStricterThanZSpec.scala:23]: does this have to be private? I suppose so, right?

**Review by BalduinLandolt — APPROVED** [review]
looks good

File: docs/02-knora-ontologies/knora-base.md [hunk 5]
```diff
@@ -32,7 +32,7 @@ In Knora, each item of data belongs to some particular project. Each project usi
 
 - `projectShortname` (1): A short name that can be used to identify the project in configuration files and the like.
 
-- `projectLongname` (1): The full name of the project.
+- `projectLongname` (0-1): The full name of the project.
```
[the remark this answers is not in the document]
Comment (irinaschubert) [reply docs/02-knora-ontologies/knora-base.md:35]: Yes, in knora-base it's defined as optional: https://github.com/dasch-swiss/dsp-api/blob/4fab193cf5db6c24a07428f19b396596d39053fe/knora-ontologies/knora-admin.ttl#L449

File: webapi/src/test/scala/org/knora/webapi/messages/v2/responder/ontologymessages/KnoraCardinalityInfoIsStricterThanZSpec.scala (hunk 4 above)
Comment (irinaschubert) [reply webapi/src/test/scala/org/knora/webapi/messages/v2/responder/ontologymessages/KnoraCardinalityInfoIsStricterThanZSpec.scala]: It doesn't have to be. As private values caused us trouble with fmt, I decided to delete it here and in other ZIO specs.

**Pull request merged**